webapp/channels/src/components/admin_console/ip_filtering/index.tsx

```diff
@@ -92,6 +115,58 @@ const IPFiltering = () => {
         }
     }, [filterToggle]);
 
+    useEffect(() => {
+        if (installationFetchAttempts > 15) {
+            // Average time for provisioner to update is around 30 seconds. This allows up to 75 seconds before it will stop fetching, displaying an error
+            setSavingDescription((
+                <>
+                    <AlertOutlineIcon size={16}/> {formatMessage({id: 'admin.ip_filtering.failed_to_fetch_installation_state', defaultMessage: 'Failed to fetch your workspaces status. Please try again later or contact support.'})}
+                </>
+            ));
+            return () => {};
+        }
+
+        const interval = setInterval(() => {
+            getInstallationStatus();
+        }, 5000);
+
+        if (installationStatus === 'stable') {
+            setSaving(false);
+            setSavingDescription(null);
+            clearInterval(interval);
+        }
+
+        return () => clearInterval(interval);
+    }, [installationStatus, installationFetchAttempts]);
```

Comment:
This effect is weird.

- We create an interval, that will either way increase the fetch attempts or the installationStatus, so we will clear the interval in the next run of the effect. So the interval will only run once, unless `getInstallationStatus` takes too long, and in that scenario you don't want to run it twice. Wouldn't make more sense a timeout?
- This effect is setting up some states, so... you are changing a state so you can re-render and set a different state so it can re-render? It does not make much sense. Most of the logic can probably be sent to the callback (`getInstallationStatus`).

Reply:
I've refactored out the useEffect in the latest commit. TL;DR:

```
 function pollInstallationStatus() {
        let installationFetchAttempts = 0;
        const interval = setInterval(async () => {
            if (installationFetchAttempts > 15) {
                // Average time for provisioner to update is around 30 seconds. This allows up to 75 seconds before it will stop fetching, displaying an error
                setSavingDescription((
                    <>
                        <AlertOutlineIcon size={16}/> {formatMessage({id: 'admin.ip_filtering.failed_to_fetch_installation_state', defaultMessage: 'Failed to fetch your workspaces status. Please try again later or contact support.'})}
                    </>
                ));
                clearInterval(interval);
                return;
            }
            const result = await dispatch(getInstallation());
            installationFetchAttempts++;
            if (result.data) {
                const {data} = result;
                if (data.state === 'stable') {
                    setSaving(false);
                    setSavingDescription(null);
                    clearInterval(interval);
                }
                setInstallationStatus(data.state);
            }
        }, 5000);
    }

    async function getInstallationStatus() {
        const result = await dispatch(getInstallation());
        if (result.data) {
            const {data} = result;
            setInstallationStatus(data.state);
            if (installationStatus === '' && data.state !== 'stable') {
                // This is the first load of the page, and the installation is not stable, so we must lock saving until it becomes stable
                setSaving(true);

                // Override the default messages for the save button and the error message to be communicative of the current state to the user
                setSavingMessage(savingButtonMessages.SAVING_PREVIOUS_CHANGE);
                changeSavingDescription(savingDescriptionMessages.SAVING_PREVIOUS_CHANGE);
            }
            if (data.state !== 'stable') {
                pollInstallationStatus();
            }
        }
    }
```

One thing to clarify - the update on the backend is happening asynchronously, so this timeout/interval isn't to account for a single long-living HTTP requests (each status request fetch should only take a few ms, it's just grabbing the workspaces state from a db column) but instead to keep checking if the async update has finished on the backend.

Reply:
I understand on the server side the process is fast, but the network is always a tricky thing :P I don't expect the network to keep stalling for 5 seconds, but who knows. Nevertheless, this version seems good :smile: